Control Explorer split to individual explorers

h-kataria opened the pull request:
In this PR, Policy Profiles accordion has been moved into it's own explorer, as part of Control explorer split. This is a pre-work for Control explorer de-explorization.
Named policy profile related files and method names to match the model name `miq_policy_set`

Cross Repo [PR](https://github.com/ManageIQ/manageiq-cross_repo-tests/pull/168)
Data Migration [PR](https://github.com/ManageIQ/manageiq-schema/pull/509)
Core [PR](https://github.com/ManageIQ/manageiq/pull/20438)

https://github.com/ManageIQ/manageiq-ui-classic/issues/6996
https://github.com/ManageIQ/manageiq-ui-classic/issues/6819

![Screenshot from 2020-08-12 10-09-13](https://user-images.githubusercontent.com/3450808/90027375-7aab5a80-dc86-11ea-8ad7-f025e9393eb5.png)

h-kataria commented:
@skateman this and related PRs are ready for review

skateman commented:
@h-kataria can you please fix the rubocop issues? And I assume the CI failures are related to the dependent PRs, right?

skateman commented on app/controllers/miq_policy_set_controller.rb:
```diff
@@ -0,0 +1,650 @@
+class MiqPolicySetController < ApplicationController
```
Why `MiqPolicySet` why not `PolicyProfile`?

h-kataria replied:
I am trying to keep all of them to match model name.

skateman replied:
wdyt @himdel, I'm a little conservative here with the amount of changes, being worried that this might break something,

h-kataria replied:
Since I am splitting all the explorers into individual controllers and adding all new views/controllers, i think this change is safe to make. As far as product features are concerned those i would have renamed some anyway, in this PR and in follow up PRs when i start de-explorizing these screens.

himdel replied:
Well, matching the model name is probably the more conservative solution, the controller name seems to get used as a rbac feature name prefix and as a fallback for table names in GTL and compare.

Then again, deliberately naming it differently would make sure there are no such hidden dependencies, that may be worth it too, and we would end up with more human friendly URLs. No idea if just renaming it would work or if other changes would be needed though.

h-kataria replied:
@himdel @skateman are we keeping it as it is now to match models or want to change the names of controller/views/product features prefixes to be more meaningful.

himdel replied:
Go with models, that's what you've been doing for the previous ones as well, right?

We can always change the URLs on the UI side, having naming consistent with models is probably more ..consistent as far as ruby code goes :)

h-kataria commented:
> @h-kataria can you please fix the rubocop issues? And I assume the CI failures are related to the dependent PRs, right?

Addressed rubocop warnings and fixed a failing spec test. Rest of the failures should be fixed once core PR is merged.

skateman commented on app/presenters/tree_builder_policy_profile.rb:
```diff
@@ -1,8 +1,4 @@
-  has_kids_for MiqPolicySet, [:x_get_tree_pp_kids]
-  has_kids_for MiqPolicy, [:x_get_tree_po_kids]
-  has_kids_for MiqEventDefinition, %i[x_get_tree_ev_kids parents]
-
```
Is this redundant from other screens? Or why are we deleting it?

skateman commented on app/controllers/miq_policy_controller.rb:
```diff
@@ -445,7 +436,7 @@ def replace_right_cell(options = {})
     replace_trees = Array(replace_trees)
     @explorer = true
 
-    trees = build_replaced_trees(replace_trees, %i[policy_profile policy condition action alert_profile alert])
```
I guess the `policy_profile` doesn't belong in this one, but in the first...right?

skateman commented on app/controllers/miq_policy_controller/conditions.rb:
```diff
@@ -1,224 +1,5 @@
 module MiqPolicyController::Conditions
   extend ActiveSupport::Concern
```
Are you keeping an empty mixin here?

skateman replied:
Nevermind, you are cleaning it up afterwards, I see it now.

skateman commented on app/controllers/mixins/policy_mixin.rb:
```diff
@@ -0,0 +1,38 @@
+module Mixins
+  module PolicyMixin
```
Does it make sense to have a mixin for a single method that is being used in a single file?

skateman commented on app/presenters/tree_builder_policy.rb:
```diff
@@ -1,6 +1,4 @@
 class TreeBuilderPolicy < TreeBuilder
-  has_kids_for MiqPolicy, [:x_get_tree_po_kids]
```
Is this visually redundant?

h-kataria replied:
@skateman yes, trees in the new explorers do not show child nodes, these trees will be gone in follow up PRs anyway. If i leave the child nodes in the trees then i will have to focus on making code changes to get the links working to jump from one explorer to another and i don't see any point in investing time getting that working as explorers/trees will be gone soon.

skateman commented on app/controllers/miq_action_controller.rb:
```diff
@@ -0,0 +1,608 @@
+
+  include Mixins::GenericSessionMixin
+  include Mixins::BreadcrumbsMixin
+  include Mixins::PolicyMixin
```
Okay, I get it now, but are you using the method defined in the previous change in this file?

skateman replied:
I just don't want to have a mixin that holds methods that are not actually shared or used multiple times.

h-kataria replied:
This mixin has methods that are being called from multiple places.

skateman commented:
Deleting a policy profile doesn't load its name in the modal:
![Screenshot from 2020-09-17 11-23-31](https://user-images.githubusercontent.com/649130/93452100-4e63a900-f8d8-11ea-8663-ccf3a1abc6cb.png)

And the flash message shows `undefined` as the name:
![Screenshot from 2020-09-17 11-23-38](https://user-images.githubusercontent.com/649130/93452135-591e3e00-f8d8-11ea-9eda-aa622bc8496b.png)

skateman commented:
The error above with the missing name also happens for policies...

When editing a policy, the paginator is shown on the bottom of the page:
![Screenshot from 2020-09-17 11-25-48](https://user-images.githubusercontent.com/649130/93452337-971b6200-f8d8-11ea-9fa6-8f376d4a190f.png)

When trying to copy a policy, it crashes:
![Screenshot from 2020-09-17 11-29-07](https://user-images.githubusercontent.com/649130/93452707-05f8bb00-f8d9-11ea-9e31-b41b98fee017.png)

h-kataria commented:
> Deleting a policy profile, a policy, an alert, an alert profile or a condition doesn't load its name in the modal:
> ![Screenshot from 2020-09-17 11-23-31](https://user-images.githubusercontent.com/649130/93452100-4e63a900-f8d8-11ea-8663-ccf3a1abc6cb.png)
> 
> And the flash message shows `undefined` as the name:
> ![Screenshot from 2020-09-17 11-23-38](https://user-images.githubusercontent.com/649130/93452135-591e3e00-f8d8-11ea-9eda-aa622bc8496b.png)

@himdel i am seeing the same issue on rebased upstream master, can you please take a look, or is there something missing in my environment

h-kataria commented:
> When editing a policy, or an alert the paginator is shown on the bottom of the page:
> ![Screenshot from 2020-09-17 11-25-48](https://user-images.githubusercontent.com/649130/93452337-971b6200-f8d8-11ea-9fa6-8f376d4a190f.png)
> 
> When trying to copy a policy, it crashes:
> ![Screenshot from 2020-09-17 11-29-07](https://user-images.githubusercontent.com/649130/93452707-05f8bb00-f8d9-11ea-9e31-b41b98fee017.png)


> When editing a policy, or an alert the paginator is shown on the bottom of the page:
> ![Screenshot from 2020-09-17 11-25-48](https://user-images.githubusercontent.com/649130/93452337-971b6200-f8d8-11ea-9fa6-8f376d4a190f.png)
> 
> When trying to copy a policy, it crashes:
> ![Screenshot from 2020-09-17 11-29-07](https://user-images.githubusercontent.com/649130/93452707-05f8bb00-f8d9-11ea-9e31-b41b98fee017.png)

@skateman fixed paging bar issue on policy edit(existing issue on master) and fixed copy policy issue. Please re-test

skateman commented:
> @skateman fixed paging bar issue on policy edit(existing issue on master) and fixed copy policy issue. Please re-test

As I wrote above, this also happens on the alerts screen

h-kataria commented:
> > @skateman fixed paging bar issue on policy edit(existing issue on master) and fixed copy policy issue. Please re-test
> 
> As I wrote above, this also happens on the alerts screen
 removed paging bar from both alert and policy edit screens

h-kataria commented:
> > Deleting a policy profile, a policy, an alert, an alert profile or a condition doesn't load its name in the modal:
> > ![Screenshot from 2020-09-17 11-23-31](https://user-images.githubusercontent.com/649130/93452100-4e63a900-f8d8-11ea-8663-ccf3a1abc6cb.png)
> > And the flash message shows `undefined` as the name:
> > ![Screenshot from 2020-09-17 11-23-38](https://user-images.githubusercontent.com/649130/93452135-591e3e00-f8d8-11ea-9eda-aa622bc8496b.png)
> 
> @himdel i am seeing the same issue on rebased upstream master, can you please take a look, or is there something missing in my environment

@skateman this issue was caused by https://github.com/ManageIQ/manageiq-ui-classic/pull/6980 on master

skateman commented:
Found this interesting issue with events, when you go to an event with assigned policies and try to click on one of these:
![Screenshot from 2020-09-23 13-46-40](https://user-images.githubusercontent.com/649130/94008406-45c11600-fda3-11ea-983a-1bdcbacd8bfd.png)

It will get you to the related page, but it crashes as the node is not existing in the left side tree:
![Screenshot from 2020-09-23 13-46-47](https://user-images.githubusercontent.com/649130/94008500-68532f00-fda3-11ea-90ac-ce997338ab94.png)

I guess this should redirect you to the policies page, so it might not be fixable in this PR but in a followup where we introduce permalinks?

h-kataria commented:
> Found this interesting issue with events, when you go to an event with assigned policies and try to click on one of these:
> ![Screenshot from 2020-09-23 13-46-40](https://user-images.githubusercontent.com/649130/94008406-45c11600-fda3-11ea-983a-1bdcbacd8bfd.png)
> 
> It will get you to the related page, but it crashes as the node is not existing in the left side tree:
> ![Screenshot from 2020-09-23 13-46-47](https://user-images.githubusercontent.com/649130/94008500-68532f00-fda3-11ea-90ac-ce997338ab94.png)
> 
> I guess this should redirect you to the policies page, so it might not be fixable in this PR but in a followup where we introduce permalinks?
> 
> On the other hand, the linking from policies to events doesn't work, there's a pointer cursor but nothing happens on a click:
> ![Screenshot from 2020-09-23 13-49-10](https://user-images.githubusercontent.com/649130/94008609-92a4ec80-fda3-11ea-8b58-c832bb4132f9.png)
> 
> Same problem also happens when you click on an Action's related policy:
> ![Screenshot from 2020-09-23 13-50-19](https://user-images.githubusercontent.com/649130/94008731-c54ee500-fda3-11ea-8edf-52231e23598c.png)
> 
> @h-kataria other than this I haven't found anything else +1

@skateman all the cross linking will be fixed with de-explorizing of these explorers.

skateman commented:
@h-kataria so are we merging this with the bug? Not complaining, just asking :thinking:

h-kataria commented:
> @h-kataria so are we merging this with the bug? Not complaining, just asking thinking

@skateman i have removed those links for now. Will add them back when de-explorizing these explorers

chessbyte commented:
@h-kataria so we are not regressing functionality with this PR?  I ask because we will be creating the Kasparov branch soon (~2 weeks), and wanted to ensure that you want this to land in the branch rather than getting it burned in first and releasing in L branch. /cc @Fryguy

h-kataria commented:
@chessbyte This PR eliminates functionality of being able to navigate to Policy related objects thru cross linking. I will be adding that functionality back with de-explorization(I didn't see any point in getting that part working when we are in process of getting rid of explorers in follow up PRs) 
In my opinion we should wait to merge this PR and have all work go in L branch, we are not gaining anything in Kasparov by merging in Control explorer split by itself.

skateman commented:
When accessing a policy from the right side table:
![Screenshot from 2020-10-29 12-43-13](https://user-images.githubusercontent.com/649130/97564153-6f91dc00-19e4-11eb-8a5d-c290843e1ec1.png)

skateman commented:
When accessing an action from the right side:
![Screenshot from 2020-10-29 12-45-06](https://user-images.githubusercontent.com/649130/97564274-9e0fb700-19e4-11eb-95e8-73bb3aca9839.png)

h-kataria commented:
> When accessing a policy from the right side table:
> ![Screenshot from 2020-10-29 12-43-13](https://user-images.githubusercontent.com/649130/97564153-6f91dc00-19e4-11eb-8a5d-c290843e1ec1.png)

i will take a look, these broke with list view conversion.

h-kataria commented:
> > When accessing a policy from the right side table:
> > ![Screenshot from 2020-10-29 12-43-13](https://user-images.githubusercontent.com/649130/97564153-6f91dc00-19e4-11eb-8a5d-c290843e1ec1.png)
> 
> i will take a look, these broke with list view conversion.

@skateman fixed reported issues

skateman commented:
On a condition's summary screen, clicking on a linked policy renders the policy's summary screen but doesn't switch the trees and throws an error:
![Screenshot from 2020-11-02 14-18-50](https://user-images.githubusercontent.com/649130/97872501-6bccc500-1d16-11eb-9e68-771b31537ce6.png)
![Screenshot from 2020-11-02 14-18-42](https://user-images.githubusercontent.com/649130/97872504-6cfdf200-1d16-11eb-8ec9-6aa570dac3a6.png)

skateman commented:
The toolbar for `Actions` is missing :confused: 

![Screenshot from 2020-11-02 14-27-19](https://user-images.githubusercontent.com/649130/97873213-9408f380-1d17-11eb-91c8-729b3c01aaf8.png)

h-kataria commented:
> The toolbar for `Actions` is missing confused
> 
> ![Screenshot from 2020-11-02 14-27-19](https://user-images.githubusercontent.com/649130/97873213-9408f380-1d17-11eb-91c8-729b3c01aaf8.png)

@skateman fixed, please re-review

skateman commented:
When trying to delete an action:
```
ActionController::RoutingError (No route matches [POST] "/miq_action/reload")
```
![Screenshot from 2020-11-02 16-11-00](https://user-images.githubusercontent.com/649130/97884184-0f71a180-1d26-11eb-962a-631bc7dc07c4.png)

h-kataria commented:
> When trying to delete an action:
> 
> ```
> ActionController::RoutingError (No route matches [POST] "/miq_action/reload")
> ```
> 
> ![Screenshot from 2020-11-02 16-11-00](https://user-images.githubusercontent.com/649130/97884184-0f71a180-1d26-11eb-962a-631bc7dc07c4.png)

@skateman fixed.

skateman reviewed (approved):
![The Seal of Approval](http://www.skateman.eu/seal.jpg)

h-kataria commented:
@skateman jest spec test failures don't look related to changes in this PR.
